lib/semantics/check-do.cc

```diff
@@ -219,6 +215,34 @@ class DoConcurrentBodyEnforce {
   }
 
 private:
+  void CheckAllocateObjectIsntCoarray(
+      const parser::AllocateObject &allocateObject, StmtType stmtType) {
+    const parser::Name &name{GetLastName(allocateObject)};
+    if (name.symbol && IsCoarray(*name.symbol)) {
+      context_.Say(name.source,
+          "%s coarray not allowed in DO CONCURRENT"_err_en_US,
+          EnumToString(stmtType));
+    }
+  }
+
+  void CheckDoesntContainCoarray(const parser::AllocateStmt &allocateStmt) {
+    const auto &allocationList{
+        std::get<std::list<parser::Allocation>>(allocateStmt.t)};
+    for (const auto &allocation : allocationList) {
+      const auto &allocateObject{
+          std::get<parser::AllocateObject>(allocation.t)};
+      CheckAllocateObjectIsntCoarray(allocateObject, StmtType::ALLOCATE);
+    }
+  }
+
+  void CheckDoesntContainCoarray(const parser::DeallocateStmt &deallocateStmt) {
+    const auto &allocateObjectList{
+        std::get<std::list<parser::AllocateObject>>(deallocateStmt.t)};
+    for (const auto &allocateObject : allocateObjectList) {
+      CheckAllocateObjectIsntCoarray(allocateObject, StmtType::DEALLOCATE);
+    }
+  }
+
   bool anyObjectIsCoarray() { return false; }  // FIXME placeholder
   bool anyObjectIsPolymorphic() { return false; }  // FIXME placeholder
   bool EndTDeallocatesCoarray() { return false; }  // FIXME placeholder
```

Comment:
Are these now dead?

Reply:
Probably.  I was planning to complete all of the semantic checking for DO constructs and then clean things up.  Some or all of these functions will probably disappear at that time.